host/vm-manager/0008-Handle-proper-allocation-of-hugepages-VFs.patch

```diff
@@ -0,0 +1,326 @@
+From f77ce8870413374d48c6e2e17c7f2b0d18ea36ba Mon Sep 17 00:00:00 2001
+From: intel <prashanth.suresh@intel.com>
+Date: Wed, 30 Aug 2023 11:32:13 +0530
+Subject: [PATCH] Handle proper allocation of hugepages, VFs
+
+- Using flocks to handle proper launch of VMs.
+
+Signed-off-by: Suresh, Prashanth <prashanth.suresh@intel.com>
+---
+ src/guest/vm_builder_qemu.cc | 147 ++++++++++++++++++++++++++++++-----
+ src/guest/vm_builder_qemu.h  |  10 +++
+ 2 files changed, 139 insertions(+), 18 deletions(-)
+
+diff --git a/src/guest/vm_builder_qemu.cc b/src/guest/vm_builder_qemu.cc
+index f1a3edc..fa9e531 100644
+--- a/src/guest/vm_builder_qemu.cc
++++ b/src/guest/vm_builder_qemu.cc
+@@ -72,6 +72,7 @@ constexpr const unsigned int kPreemptTimeout = 50000;
+ constexpr const unsigned int kExecQuantum = 25;
+ 
+ constexpr const char *kQmpPowerSocket = "/tmp/qmp-pwr-socket-";
++constexpr const char *kSetupLock = "/tmp/sriov.setup.lock";
+ 
+ const std::string connectors[8] = {
+                 "HDMI-1",
+@@ -193,15 +194,9 @@ static bool SetupHugePages(const std::string &mem_size) {
+         return false;
+     }
+ 
+-    /* Get free Huge pages */
+-    int free_hp = ReadSysFile(kSys2MFreeHugePages, std::ios_base::dec);
+-
+-    if (free_hp >= mem_mb/2)
+-        return true;
+-
+     /* Get nr hugepages */
+     int nr_hp = ReadSysFile(kSys2MNrHugePages, std::ios_base::dec);
+-    int required_hp = nr_hp - free_hp + mem_mb/2;
++    int required_hp = nr_hp + mem_mb/2;
+ 
+     /* Try to allocate required huge pages */
+     WriteSysFile(kSys2MNrHugePages, std::to_string(required_hp));
+@@ -222,6 +217,40 @@ static bool SetupHugePages(const std::string &mem_size) {
+     return true;
+ }
+ 
++void VmBuilderQemu::CleanupHugePages(void) {
++    LOG(info) << "Cleaning up huge pages";
++    /* Get nr hugepages */
++    int nr_hp = ReadSysFile(kSys2MNrHugePages, std::ios_base::dec);
++    int new_nr_hp = nr_hp;
++    int hp_alloc = 0;
```

Comment:
hp_alloc should not be a local variable set to 0.
It should contain the value of allocated hugepage upon allocation in SetupHugePages().

For example, we have a situation of 4 Vms running with the total hugepage at nr_hp
When we quit 1 of them, we should deallocate the hugepage for that one VM.
Thus the new hugepage should be "new_nr_hp = nr_hp - hp_alloc", where hp_alloc is the hugepage to be freed.
The variable hp_alloc should keep track of the hugepage allocated for a particular VM in SetupHugePages().